code/modules/shuttle/emergency.dm

```diff
@@ -6,61 +11,164 @@
 	var/auth_need = 3
 	var/list/authorized = list()
 
+/obj/machinery/computer/emergency_shuttle/attackby(obj/item/I, mob/user,params)
+	if(istype(I, /obj/item/weapon/card/id))
+		say("Please equip your ID card into your ID slot to authenticate.")
+	. = ..()
 
-/obj/machinery/computer/emergency_shuttle/attackby(obj/item/I, mob/user, params)
-	var/obj/item/weapon/card/id/ID = I.GetID()
-	if(ID)
-		if(stat & (BROKEN|NOPOWER))
-			return
-		if(SSshuttle.emergency.mode != SHUTTLE_DOCKED)
-			return
-		if(SSshuttle.emergency.timeLeft() < 11)
-			return
-		if(!(access_heads in ID.access)) //doesn't have this access
-			user << "The access level of [ID.registered_name]\'s card is not high enough. "
-			return
+/obj/machinery/computer/emergency_shuttle/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, datum/tgui/master_ui = null, datum/ui_state/state = human_adjacent_state)
+
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "emergency_shuttle_console", name,
+			400, 400, master_ui, state)
+		ui.open()
+
+/obj/machinery/computer/emergency_shuttle/ui_data()
+	var/list/data = list()
+
+	data["timer_str"] = SSshuttle.emergency.getTimerStr()
+	data["engines_started"] = ENGINES_STARTED
+	data["authorizations_remaining"] = max((auth_need - authorized.len), 0)
+	var/A = list()
+	data["authorizations"] = A
```

Comment:
this weirds me out so much.
I know lists are objects.
I know they're pass by reference.
but you're setting data before it's been added and it really just irks me.